crates/volta-migrate/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+use volta_core::layout::volta_home;
+#[cfg(unix)]
+use volta_core::layout::volta_install;
+use volta_core::shim;
+use volta_fail::{Fallible, ResultExt};
+
+enum MigrationState {
+    Empty(empty::Empty),
+    V0(Box<V0>),
+    V1(Box<V1>),
+}
+
+impl MigrationState {
+    fn current() -> Fallible<Self> {
+        /*
+        Triage for determining the current layout version:
+        - Does the V1 layout file exist? If yes then V1
+        - Does Volta Home exist? If yes (Windows) then V0
+            - If yes (Unix) then check if Volta Install is outside shim_dir?
+                - If yes, then V0
+                - If no, then check if $VOLTA_HOME/load.sh exists? If yes then V0
+        - Else Empty
+
+        The extra logic on Unix is necessary because Unix installs can be either inside or outside $VOLTA_HOME
+        If it is inside, then the directory necessarily must exist, so we can't use that as a determination.
+        If it is outside (and for Windows which is always outside), then if $VOLTA_HOME exists, it must be from a
+        previous, V0 installation.
+
+        Going forward, each new version will have an associated layout file, so we can use that as a discriminant
+        */
+
+        let home = volta_home()?;
+        let volta_home = home.root().to_owned();
+
+        if home.layout_file().exists() {
+            return Ok(MigrationState::V1(Box::new(V1::new(volta_home))));
+        }
+
+        if volta_home.exists() {
+            #[cfg(windows)]
+            return Ok(MigrationState::V0(Box::new(V0::new(volta_home))));
+
+            #[cfg(unix)]
+            {
+                let install = volta_install()?;
+                if install.root().starts_with(&volta_home) {
+                    // Installed inside $VOLTA_HOME, so need to look for `load.sh` as a marker
+                    if volta_home.join("load.sh").exists() {
+                        return Ok(MigrationState::V0(Box::new(V0::new(volta_home))));
+                    }
+                } else {
+                    // Installed outside of $VOLTA_HOME, so it must exist from a previous V0 install
+                    return Ok(MigrationState::V0(Box::new(V0::new(volta_home))));
+                }
+            }
+        }
+
+        Ok(MigrationState::Empty(empty::Empty::new(volta_home)))
+    }
+}
+
+pub fn run_migration() -> Fallible<()> {
+    let mut state = MigrationState::current()?;
+
+    loop {
+        state = match state {
+            MigrationState::Empty(e) => MigrationState::V1(Box::new(e.try_into()?)),
+            MigrationState::V0(zero) => MigrationState::V1(Box::new((*zero).try_into()?)),
+            MigrationState::V1(one) => {
+                one.commit()?;
+                break;
+            }
+        };
+    }
+
+    Ok(())
+}
+
+fn regenerate_shims_for_dir(dir: &Path) -> Fallible<()> {
+    debug!("Rebuilding shims");
+    for shim_name in get_shim_list_deduped(dir)?.iter() {
+        shim::delete(shim_name)?;
```

Comment:
I don't quite see how deleting and recreating makes it possible to gracefully restart. Is the idea that anything you delete is recreate-able? I'm just a little nervous you could delete some data somewhere along the line that you can't recreate.

Would it be safer to recreate/copy everything fresh to a temp directory and then have an atomic (or almost-atomic, anyway) directory swap at the end?

Reply:
That's a good point that this isn't going support graceful restarting. Since the shims are symlinks, the only way (I believe) to re-target them is to remove and recreate them. For these specifically, I think we can probably do a temp directory version, since we can read through the existing directory, create whatever contents / shims we need in the temp, and then swap them over.

I initially wanted this (and the writing of the layout file) to be separated from individual migrations since it would potentially be duplicating work to have them run with each "step". However, thinking more deeply, we actually do want these operations to happen at each migration, because that lets the graceful restart actually apply happen: Any migrations that have completed will be skipped on subsequent restarts, leaving only the remaining work. It also prevents us from missing a regeneration of the shims or having to include it in every migration.

Reply:
Letting this percolate overnight and looking at the code this morning, I'm actually not sure there is a more "restartable" solution. The issue with a temporary directory is that if we get a failure during removing the existing directory or while `rename`ing the temp directory, we lose _all_ the contents of the shim directory (which could also include the volta binaries themselves). While this approach of removing and re-creating the shims has more potential points of failure, each failure will only affect a single shim, as opposed to the entire directory. This means at worst (if we successfully delete a shim but then fail to re-create it), we lose information about a single shim, and the error should give information about which shim failed.

It won't be an easy thing to recover from, but it _will_ be easier than if we failed the operation in a way that lost all of the information about the shim directory at once.